tensorizer/stream_io.py

```diff
@@ -417,6 +422,292 @@ def seek(self, position, whence=SEEK_SET):
+        self._redis = redis.Redis(host=host, port=port, db=0)
+        self._redis_tcp = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self._redis_tcp.setsockopt(
+            socket.SOL_SOCKET, socket.SO_RCVBUF, buffer_size
+        )
+        self._redis_tcp.connect((host, port))
+        self._redis_tcp_mutex = threading.Lock()
+
+        # Do a key scan for the prefix, and collect all the indexes from the keys.
+        self._keys = []
+        self._indexes = []
+        self._sizes = []
+        keys = self._redis.scan_iter(f"{prefix}:*")
+        # Sort the keys by their index.
+        for key in sorted(
+            keys, key=lambda x: int(x.decode("utf-8").split(":")[-1])
+        ):
+            self._keys.append(key)
+
+        # Get indexes.
+        largest = 0
+        for key in self._keys:
+            size = self._redis.strlen(key)
+            if size > largest:
+                largest = size
+            self._sizes.append(size)
+            self._indexes.append(int(key.decode("utf-8").split(":")[-1]))
+
+        self._curr = 0
+        self._curr_key = 0
+        self._curr_buffer = bytearray(largest)
+        self._curr_buffer_idx = -1
+        self._curr_buffer_view = memoryview(self._curr_buffer)[0:0]
+        self.closed = False
+
+        init_end = time.monotonic()
+
+        self.setup_latency = init_end - init_begin
+        self.response_latencies = []
+        self.bytes_read = 0
+        self.bytes_skipped = 0
+        self.read_operations = 0
+
+    def _find_key_index(self, position):
+        for i, index in enumerate(self._indexes):
+            if position < index:
+                return i - 1
+        return len(self._keys) - 1
+
+    def _read_from(
+        self,
+        position: int,
+        size: int,
+        ba: Union[bytearray, memoryview],
+        no_buffer: bool = False,
+    ) -> int:
+        self.read_operations += 1
+        curr_key = self._find_key_index(position)
+        curr_key_pos = self._indexes[curr_key]
+        curr_key_sz = self._sizes[curr_key]
+        begin_idx = position - curr_key_pos
+        position_end = position + size
+        read_sz = min(size, curr_key_sz)
+        curr_buffer_end = self._curr_buffer_idx + len(self._curr_buffer_view)
+
+        # Check if we have a partial read of the current key in the buffer.
+        # TODO: allow for reads that do not align with the key boundaries.
+        if (
+            not no_buffer
+            and len(self._curr_buffer_view) > 0
+            and self._curr_buffer_idx
+            <= position
+            < position_end
+            < curr_buffer_end
+        ):
+            num_bytes = min(read_sz, len(self._curr_buffer_view))
+            ba[:num_bytes] = self._curr_buffer_view[:num_bytes]
+            self._curr_buffer_idx += num_bytes
+            self._curr_buffer_view = self._curr_buffer_view[num_bytes:]
+            return num_bytes
+
+        self._redis_tcp_mutex.acquire(True)
```

Comment:
I'd suggest dropping the mutex in this code—but if it's kept, it should use the `with self._redis_tcp_mutex:` form for more exception safety

Reply:
Addressed.